### PR TITLE
Add applications count and admin name to GET /accounts response

### PR DESCRIPTION
**Pull request by josemigallas**
**What this PR does / why we need it**:

For Portafly's Accounts Listing we need 2 more fields:
- The count of applications for each account
- The name of the admin

After adding the new fields in the response, we need to update the DAL in portafly to match the same name (`apps_count` and `admin_name`).

Tasks:
- [x] Add a new column to Accounts for the counter: #1967
- [x] Back fill Accounts using a rake task
- [x] Add bought_cinstances_count to the response
- [ ] Tests in spec/acceptance/api/account_spec.rb
- [x] Tests in test/unit/tasks
- [x] Add admin_name field to the response
- [x] Update DAL in Portafly to include the new fields (verify it renders in Dev Accounts Table)
- [x] Update Account index in Porta (and other places probably) to benefit from the new column

**Which issue(s) this PR fixes** 

[THREESCALE-5328: Get the applications count for each account](https://issues.redhat.com/browse/THREESCALE-5328)

**Verification steps:**
Try it locally with curl:
```
$ curl -v  -X GET "http://provider-admin.example.com.lvh.me:3000/admin/api/accounts.json?access_token=<TOKEN>
```
The response must contain the new fields.

### CODE REVIEW

File: app/representers/account_representer.rb [hunk 1]
```diff
@@ -18,6 +18,7 @@ module AccountRepresenter
   end
 
   with_options(if: ->(*) { provider? }) do
+    # TODO: property :admin_name
```
Comment (josemigallas) [inline app/representers/account_representer.rb:21]: This will go in the next PR

File: app/representers/account_representer.rb [hunk 2]
```diff
@@ -54,6 +55,8 @@ module AccountRepresenter
   end
 
   property :state
+  property :bought_cinstances_count
+  property :admin_name
```
Comment (Martouta) [inline app/representers/account_representer.rb:59]:
```suggestion
  property :admin_user_name
```
WDYT? because `admin` is not clear that it is refering to `admin user`, specially when there are account attributes called `admin_domain` and `admin_base_url`

Comment (Martouta) [inline app/representers/account_representer.rb:59]: These 2 need to be tested in:
https://github.com/3scale/porta/blob/23c541044bf72ea2e9632129877ef0d393e9e32c/spec/acceptance/api/account_spec.rb#L5

File: db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+class AddBoughtCinstancesCountToAccounts < ActiveRecord::Migration[5.0]
```
Comment (Martouta) [inline db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb:1]: This needs to be migrated for the 3 schemas 😝

File: app/representers/account_representer.rb (hunk 2 above)
Comment (josemigallas) [reply app/representers/account_representer.rb:59]: The problem with `admin_user_name` is that it can be confused with the admin's `user_name`. It could be `admin_display_name`?

File: db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb (hunk 3 above)
Comment (Martouta) [reply db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb:1]: And please ensure that it works for the 3 DBs also when it contains data for the affected tables/columns

File: app/representers/account_representer.rb (hunk 2 above)
Comment (Martouta) [reply app/representers/account_representer.rb:59]: @josemigallas yes that's good enough I guess

File: app/representers/account_representer.rb [hunk 4]
```diff
@@ -67,5 +70,9 @@ def credit_card_stored
     credit_card_stored?
   end
 
+  def admin_name
+    account.admins.first.try(:display_name)
```
Comment (Martouta) [inline app/representers/account_representer.rb:74]: You may want to use one of these methods instead of doing `.admins.first` here 😄 
https://github.com/3scale/porta/blob/23c541044bf72ea2e9632129877ef0d393e9e32c/app/models/account.rb#L192-L198

Comment (Martouta) [inline app/representers/account_representer.rb:74]: `account.` is not necessary 😄 you are kinda already inside the Account 😉 
So it would be just `first_admin.try!(:display_name)`, but it is even better if you delegate IMO 😄 
like:
```
delegate :display_name, to: :first_admin, allow_nil: true
alias admin_display_name display_name
```

File: db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  def change
+    safety_assured do
+      add_column :accounts, :bought_cinstances_count, :integer, :null => false, :default => 0
```
Comment (Martouta) [inline db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb:4]: This is a blocking migration so how do we plan to run this for SaaS?

Comment (hallelujah) [reply db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb:4]: And we should not add anymore column to accounts, this table is terribly huge and is used every time

File: app/models/contract.rb [hunk 6]
```diff
@@ -39,7 +39,7 @@ class Contract < ApplicationRecord
   validates :user_key, length: { maximum: 256 }
 
   # TODO: rename to buyer_account and remove alias
-  belongs_to :user_account, class_name: 'Account', autosave: false
+  belongs_to :user_account, class_name: 'Account', autosave: false, counter_cache: true
```
Comment (Martouta) [inline app/models/contract.rb:42]:
```suggestion
  belongs_to :user_account, class_name: 'Account', autosave: false, counter_cache: :bought_cinstances_count
```

File: db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb (hunk 5 above)
Comment (macejmic) [reply db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb:4]: @hallelujah and where you would create that counter attribute?

File: lib/tasks/accounts.rake [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+namespace :accounts do
+
+  desc "Count bought cinstnaces for each account"
```
Comment (Martouta) [inline lib/tasks/accounts.rake:5]:
```suggestion
  desc "Reset the counter of bought_cinstances for all accounts"
```

File: lib/tasks/accounts.rake [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  task count_bought_cinstances: :environment do
```
Comment (Martouta) [inline lib/tasks/accounts.rake:6]:
```suggestion
  task reset_bought_cinstances_count: :environment do
```
or:
```ruby
  task reset_bought_cinstances_counter: :environment do
```

Comment (Martouta) [inline lib/tasks/accounts.rake:6]: Add a test in [/test/unit/tasks/...](https://github.com/3scale/porta/tree/master/test/unit/tasks)

File: test/unit/tasks/accounts_test.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+module Tasks
+  class AccountsTest < ActiveSupport::TestCase
+    test 'accounts:reset_bought_cinstances_count' do
+      account = FactoryBot.create(:account)
+      account.update(bought_cinstances_count: nil)
```
Comment (Martouta) [inline test/unit/tasks/accounts_test.rb:9]: This works:
```ruby
# frozen_string_literal: true

require 'test_helper'

module Tasks
  class AccountsTest < ActiveSupport::TestCase
    test 'accounts:reset_bought_cinstances_count' do
      account = FactoryBot.create(:simple_buyer)
      FactoryBot.create(:cinstance, user_account: account)

      sql_update_nil = <<-SQL
        UPDATE  accounts
        SET     bought_cinstances_count = NULL
        WHERE   accounts.id = #{account.id}
      SQL
      ActiveRecord::Base.connection.execute(sql_update_nil)

      assert_nil account.reload.bought_cinstances_count

      execute_rake_task 'accounts.rake', 'accounts:reset_bought_cinstances_count'

      account.reload
      assert_equal account.bought_cinstances_count, account.bought_cinstances.count
    end
  end
end
```

File: lib/tasks/accounts.rake [hunk 10]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+require 'progress_counter'
+
+namespace :accounts do
+
+  desc "Reset the counter of bought_cinstances for all accounts"
+  task reset_bought_cinstances_count: :environment do
+    accounts = Account.where.not(state: :scheduled_for_deletion)
```
Comment (Martouta) [inline lib/tasks/accounts.rake:9]:
```suggestion
    accounts = Account.without_deleted
```
Using:
https://github.com/3scale/porta/blob/26657694321c1d922879efc69b033d8c9c6b1e9f/app/models/account/states.rb#L96

File: app/models/account.rb [hunk 11]
```diff
@@ -102,8 +102,13 @@ def self.columns
 
   before_validation(on: :create, if: :provider?) { generate_s3_prefix }
   before_validation(on: :create, if: :provider?) { generate_domains }
+  before_validation(on: :create) { default_bought_cinstances_count }
```
Comment (Martouta) [inline app/models/account.rb:105]: This scenario should have a test too 😄

File: app/views/buyers/accounts/_search_results.html.slim [hunk 12]
```diff
@@ -50,7 +50,7 @@ table class="data" id="buyer_accounts"
           td class="plan" = h account.bought_account_plan.try(:name)
 
         - if can? :see, :multiple_applications
-          td = link_to (current_user.accessible_cinstances.bought_by(account).count), admin_buyers_account_applications_path(account)
+          td = link_to account.bought_cinstances_count, admin_buyers_account_applications_path(account)
```
Comment (Martouta) [inline app/views/buyers/accounts/_search_results.html.slim:53]: I am not sure if it is the same 🤔 `account.bought_cinstances_count` is the number of bought cinstances. `current_user.accessible_cinstances.bought_by(account).count)` is the number of bought **accessible** cinstances. What I do not know is a non-accessible cinstances can be bought at any point in time.

File: test/unit/account_test.rb [hunk 13]
```diff
@@ -864,4 +864,23 @@ def test_smart_destroy_master
     assert master_account.reload
     refute master_account.reload.scheduled_for_deletion?
   end
+
+  test 'bought_cinstances_count when buying a cinstance' do
+    account = FactoryBot.create(:simple_buyer, provider_account: master_account)
+
+    assert_difference -> { account.bought_cinstances_count }, 1 do
+      account.buy!(master_account.default_service.published_plans.first)
```
Comment (Martouta) [inline test/unit/account_test.rb:872]:
```suggestion
      account.buy!(master_account.default_service.published_plans.first!)
```

File: test/unit/account_test.rb [hunk 14]
```diff
@@ -864,4 +864,23 @@ def test_smart_destroy_master
     assert master_account.reload
     refute master_account.reload.scheduled_for_deletion?
   end
+
+  test 'bought_cinstances_count when buying a cinstance' do
+    account = FactoryBot.create(:simple_buyer, provider_account: master_account)
```
Comment (Martouta) [inline test/unit/account_test.rb:869]:
```suggestion
    account = FactoryBot.create(:simple_provider, provider_account: master_account)
```
Because a tenant is a buyer of master which is what you are doing 🤔

File: test/unit/account_test.rb [hunk 15]
```diff
@@ -864,4 +864,23 @@ def test_smart_destroy_master
...
+
+    assert_difference -> { account.bought_cinstances_count }, 1 do
+      account.buy!(master_account.default_service.published_plans.first)
+      account.reload.bought_cinstances_count
+    end
+  end
+
+  test 'bought_cinstances_count when deleting a cinstance' do
+    account = FactoryBot.create(:simple_buyer, provider_account: master_account)
+    cinstance = account.buy!(master_account.default_service.published_plans.first)
```
Comment (Martouta) [inline test/unit/account_test.rb:879]: same as https://github.com/3scale/porta/pull/1939/files#r442742867

File: test/unit/account_test.rb [hunk 16]
```diff
@@ -864,4 +864,23 @@ def test_smart_destroy_master
...
+    account = FactoryBot.create(:simple_buyer, provider_account: master_account)
+
+    assert_difference -> { account.bought_cinstances_count }, 1 do
+      account.buy!(master_account.default_service.published_plans.first)
+      account.reload.bought_cinstances_count
+    end
+  end
+
+  test 'bought_cinstances_count when deleting a cinstance' do
+    account = FactoryBot.create(:simple_buyer, provider_account: master_account)
```
Comment (Martouta) [inline test/unit/account_test.rb:878]: same as https://github.com/3scale/porta/pull/1939/files#r442743238

File: db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb (hunk 5 above)
Comment (Martouta) [reply db/migrate/20200615130642_add_bought_cinstances_count_to_accounts.rb]: About this: we decided that now 'arbitrarily' splitting the accounts table by this criteria of just being a new attribute will just make the problem worse, so we should add this column to accounts + make a Jira issue to split the accounts table under a well-thought criteria if we end up doing so.

File: app/views/buyers/accounts/_search_results.html.slim (hunk 12 above)
Comment (josemigallas) [reply app/views/buyers/accounts/_search_results.html.slim:53]: That's a bit worrying because the whole point of the new column is to provide Portafly with this very value.

Comment (Martouta) [reply app/views/buyers/accounts/_search_results.html.slim:53]: I think that it is correct but I am not 100% sure so I left the comment in case that someone can be sure of that 😄 
I believe that only accessible plans can be bought, what I do not know for sure if a plan that has been accessible, and has bought plans can become inaccessible later on. I hope not for our mental health 😛

File: app/lib/backend/storage_rewrite.rb [hunk 17]
```diff
@@ -63,7 +63,7 @@ def rewrite_provider(id)
       metrics = provider.metrics.includes(:service, :parent)
       usage_limits = provider.usage_limits.includes(plan: :service)
 
-      total_count = services.size + cinstances.size + metrics.size + usage_limits.size + bought_cinstances.size
+      total_count = services.size + cinstances.size + metrics.size + usage_limits.size + provider.bought_cinstances_count
```
Comment (hallelujah) [inline app/lib/backend/storage_rewrite.rb:66]: I wonder, `services.size`, `cinstances.size` etc ... they are loading the records but `bought_cinstances_count` is not loading `bought_cinstances`

Why this change? Don't we need `bought_cinstances` later on? If we do then I think there is no need to use the cached counter

File: app/models/contract.rb [hunk 18]
```diff
@@ -296,7 +296,7 @@ def change_plan_internal(new_plan, &block)
   private
 
   def reset_counter_cache_for
-    [:plan].freeze
+    [:plan, :account].freeze
```
Comment (Martouta) [inline app/models/contract.rb:299]: First, RuboMarta strikes 😝 :
```suggestion
    %i[plan account].freeze
```
Now my "questions/suggestions":
1. Is it `account` or `user_account` ? why?
2. All this needs a test 😄 
3. Shouldn't the next method (`update_counter_cache?`) have this same association in the first `when` ? if not, why? because I believe that it shouldn't update this cache if the account is being deleted 😕

File: app/lib/backend/storage_rewrite.rb (hunk 17 above)
Comment (hallelujah) [reply app/lib/backend/storage_rewrite.rb:66]: Also the counter cache is made so you do not have to use `#bought_cinstances_count` but plainly do `bought_cinstances.size`

> The method “size” will use the counter_cache if its present, where as using “.count” itself would do the actual sql count.

File: app/views/buyers/accounts/_search_results.html.slim (hunk 12 above)
Comment (guicassolato) [reply app/views/buyers/accounts/_search_results.html.slim:53]: Strictly, it's not the same. The access is granted to a user for a given service (or set of services), not for buyer accounts or applications. Consider the following:
- current_user has access to services A and B, but not C
- account X bought cinstances of A and C (1 of each)
- `current_user.accessible_cinstances.bought_by(account_x).count` should be 1
- `account_x. bought_cinstances_count` will return 2

File: app/models/contract.rb (hunk 18 above)
Comment (josemigallas) [reply app/models/contract.rb:299]: @macejmic

File: app/views/buyers/applications/_widget.html.erb [hunk 19]
```diff
@@ -1,5 +1,4 @@
-<%- cinstances = @account.bought_cinstances -%>
-<%- cinstance = cinstances.first! if cinstances.count == 1 -%>
+<%- cinstance = @account.bought_cinstances.first! if @account.bought_cinstances_count == 1 -%>
```
Comment (Martouta) [inline app/views/buyers/applications/_widget.html.erb:1]: I wonder why not `<%- cinstance = @account.bought_cinstances.first -%>` 🤔 because I guess that internally the DB should use the index we have on cinstances user_account_id and it would quickly see that it doesn't have any 🤔

File: app/models/account.rb [hunk 20]
```diff
@@ -576,6 +581,10 @@ def independent_mapping_rules_enabled?
 
   private
 
+  def reset_counter_cache
+    update_column(:bought_cinstances_count, bought_cinstances.size) if persisted?
```
Comment (Martouta) [inline app/models/account.rb:585]: why is this necessary? 😕 shouldn't it do this automatically without overriding this method?

Comment (josemigallas) [reply app/models/account.rb:585]: @macejmic

File: app/views/buyers/applications/_widget.html.erb (hunk 19 above)
Comment (josemigallas) [reply app/views/buyers/applications/_widget.html.erb:1]: We should ask @macejmic 
https://github.com/3scale/system/blame/master/app/views/buyers/applications/_widget.html.erb

Comment (macejmic) [reply app/views/buyers/applications/_widget.html.erb:1]: `<%- cinstance = @account.bought_cinstances.first -%>` looks good to me.

File: app/models/account.rb (hunk 20 above)
Comment (macejmic) [reply app/models/account.rb:585]: `counter_cache` is not defined for `bought_cinstances` so yes, it has to be done manually, that's the concept of the manual counter cache.

File: app/models/contract.rb (hunk 18 above)
Comment (macejmic) [reply app/models/contract.rb:299]: `alias account user_account` so it does not matter, if it's `account` or `user_account`.

Comment (macejmic) [reply app/models/contract.rb:299]: 3. https://github.com/3scale/porta/blob/master/app/models/contract.rb#L302

so the counter is not being updated when the provider account is being deleted.

Comment (macejmic) [reply app/models/contract.rb:299]: 2. this is an "integration" unit test that covers the main use case:

```
test 'bought_cinstances_count when deleting a cinstance' do
    account = FactoryBot.create(:simple_provider, provider_account: master_account)
    cinstance = account.buy!(master_account.default_service.published_plans.first!)

    assert_difference -> { account.reload.bought_cinstances_count }, -1 do
      cinstance.destroy
      account.reload.bought_cinstances_count
    end
  end
```

all other are already covered..

**Comment by josemigallas** [comment]
Apps no longer part of this screen so it's reduced to admin_name. I'll leave this branch for future reference but close the PR in favor of #1976
